lib/src/core/authentication.ts

```diff
@@ -0,0 +1,220 @@
+ * under the License.
+ */
+
+import { AsgardeoAuthClient, AuthClientConfig, CryptoUtils, Store } from "@asgardeo/auth-js";
+import { AsgardeoAuthException } from "../exception";
+import { AuthURLCallback, NodeTokenResponse } from "../models";
+import { UserSession } from "../session";
+import { MemoryCacheStore } from "../stores";
+import { NodeCryptoUtils } from "../utils/crypto-utils";
+
+export class AsgardeoNodeCore<T>{
+
+    private _auth: AsgardeoAuthClient<T>;
+    private _cryptoUtils: CryptoUtils;
+    private _store: Store;
+    private _sessionStore: UserSession;
+
+    constructor(config: AuthClientConfig<T>, store?: Store) {
+
+        //Initialize the default memory cache store if an external store is not passed.
+        if (!store) {
+            this._store = new MemoryCacheStore();
+        } else {
+            this._store = store;
+        }
+        this._cryptoUtils = new NodeCryptoUtils();
+        this._auth = new AsgardeoAuthClient(this._store, this._cryptoUtils);
+        this._sessionStore = new UserSession(this._store);
+        this._auth.initialize(config);
+    }
+
+    public async signIn(authURLCallback: AuthURLCallback, authorizationCode?: string, sessionState?: string)
+        : Promise<NodeTokenResponse> {
+
+        //Check if the authorization code or session state is there.
+        //If so, generate the access token, otherwise generate the auth URL and return with callback function.
+
+        if (!authorizationCode || !sessionState) {
+            const authURL = await this.getAuthURL();
+            authURLCallback(authURL);
+
+            return Promise.resolve({
+                accessToken: "",
+                expiresIn: "",
+                idToken: "",
+                refreshToken: "",
+                scope: "",
+                session: "",
+                tokenType: ""
+            })
+        } else {
+            const tokenResponse = await this.requestAccessToken(authorizationCode, sessionState);
+            if (tokenResponse) {
+                return Promise.resolve(tokenResponse);
+            }
+        }
+
+        return Promise.reject(
+            new AsgardeoAuthException(
+                "AUTH_CORE-RAT1-NF01", //TODO: Not sure
+                "node-authentication",
+                "signIn",
+                "Access token or decoded token failed.",
+                "No token endpoint was found in the OIDC provider meta data returned by the well-known endpoint " +
+                "or the token endpoint passed to the SDK is empty."
+            )
+        );
+    }
+
+    public async getAuthURL(): Promise<string> {
+
+        const authURL = await this._auth.getAuthorizationURL();
+
+        if (authURL) {
+            return Promise.resolve(authURL.toString())
+        } else {
+            return Promise.reject(
+                new AsgardeoAuthException(
+                    "AUTH_CORE-RAT1-NF01", //TODO: Not sure
```

Comment:
There are some TODOs in the file, can we check whether those are resolved now ?

Reply:
This has been fixed with https://github.com/asgardeo/asgardeo-auth-node-sdk/commit/b9f0e10f9950e726e756c16f12db64be6a8dc2ea